flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java

```diff
@@ -0,0 +1,571 @@
+            ImmutableIntList inputIndices = getInputIndices(projects, buildIndices);
+            if (inputIndices.isEmpty()) {
+                return Optional.empty();
+            }
+            return findSuitableBuildSide(calc.getInput(), inputIndices, buildSideChecker);
+
+        } else if (rel instanceof Join) {
+            Join join = (Join) rel;
+            if (!(join.getLeft() instanceof Exchange) && !(join.getRight() instanceof Exchange)) {
+                // TODO: Is there such case?
+                return Optional.empty();
+            }
+
+            Tuple2<ImmutableIntList, ImmutableIntList> tuple2 = getInputIndices(join, buildIndices);
+            ImmutableIntList leftIndices = tuple2.f0;
+            ImmutableIntList rightIndices = tuple2.f1;
+
+            if (leftIndices.isEmpty() && rightIndices.isEmpty()) {
+                return Optional.empty();
+            }
+
+            boolean firstCheckLeft = !leftIndices.isEmpty() && join.getLeft() instanceof Exchange;
+            Optional<BuildSideInfo> buildSideInfoOpt = Optional.empty();
+            if (firstCheckLeft) {
+                buildSideInfoOpt =
+                        findSuitableBuildSide(join.getLeft(), leftIndices, buildSideChecker);
+                if (!buildSideInfoOpt.isPresent() && !rightIndices.isEmpty()) {
+                    buildSideInfoOpt =
+                            findSuitableBuildSide(join.getRight(), rightIndices, buildSideChecker);
+                }
+                return buildSideInfoOpt;
+            } else {
+                if (!rightIndices.isEmpty()) {
+                    buildSideInfoOpt =
+                            findSuitableBuildSide(join.getRight(), rightIndices, buildSideChecker);
+                    if (!buildSideInfoOpt.isPresent() && !leftIndices.isEmpty()) {
+                        buildSideInfoOpt =
+                                findSuitableBuildSide(
+                                        join.getLeft(), leftIndices, buildSideChecker);
+                    }
+                }
+            }
+            return buildSideInfoOpt;
+        } else if (rel instanceof BatchPhysicalGroupAggregateBase) {
+            BatchPhysicalGroupAggregateBase agg = (BatchPhysicalGroupAggregateBase) rel;
+            int[] grouping = agg.grouping();
+
+            // If one of keys are aggregate function field, return directly.
+            for (int k : buildIndices) {
+                if (k >= grouping.length) {
+                    return Optional.empty();
+                }
+            }
+
+            return findSuitableBuildSide(
+                    agg.getInput(),
+                    ImmutableIntList.copyOf(
+                            buildIndices.stream()
+                                    .map(index -> agg.grouping()[index])
+                                    .collect(Collectors.toList())),
+                    buildSideChecker);
+
+        } else {
+            // more cases
```

Comment:
We need to verify if we have at least covered all the TPC-DS cases that can apply runtime filters.

Reply:
TPCDS-related tests are still in progress